File: CodingCamp-20Apr26-RizkiPutraRamadhan-main/tests/pure-functions.js
/**
 * Pure functions extracted from js/app.js for testing.
 * These must stay in sync with the implementations in js/app.js.
 *
 * All functions here are pure (no DOM, no localStorage, no side effects).
 */

// ============================================================
// --- Theme controller (pure logic) ---
// ============================================================

/**
 * Return the opposite theme.
 * toggleTheme(toggleTheme(theme)) === theme  (involution)
 *
 * @param {"light"|"dark"} theme
 * @returns {"light"|"dark"}
 */
export function toggleThemePure(theme) {
  return theme === "light" ? "dark" : "light";
}

// ============================================================
// --- Greeting widget ---
// ============================================================

/**
 * Map an hour (0–23) to the appropriate time-of-day greeting.
 *
 * @param {number} hour - Hour in 24-hour format (0–23)
 * @returns {string}
 */
export function getGreetingPhrase(hour) {
  if (hour >= 5 && hour <= 11) {
    return "Good Morning";
  } else if (hour >= 12 && hour <= 17) {
    return "Good Afternoon";
  } else if (hour >= 18 && hour <= 21) {
    return "Good Evening";
  } else {
    return "Good Night";
  }
}

// ============================================================
// --- Timer widget ---
// ============================================================

/**
 * Enum-like object representing the three possible timer states.
 */
export const TimerState = {
  IDLE:    "IDLE",
  RUNNING: "RUNNING",
  PAUSED:  "PAUSED",
};

/**
 * Apply a timer action to a given state and return the new state.
 * This is the pure state-machine transition function.
 *
 * Transition table:
 *   IDLE    + start → RUNNING
 *   RUNNING + stop  → PAUSED
 *   RUNNING + reset → IDLE
 *   PAUSED  + start → RUNNING
 *   PAUSED  + reset → IDLE
 *   IDLE    + stop  → IDLE   (no-op)
 *   IDLE    + reset → IDLE   (no-op)
 *   PAUSED  + stop  → PAUSED (no-op, not in spec but safe default)
 *
 * @param {"IDLE"|"RUNNING"|"PAUSED"} state
 * @param {"start"|"stop"|"reset"} action
 * @returns {"IDLE"|"RUNNING"|"PAUSED"}
 */
export function applyTimerAction(state, action) {
  switch (action) {
    case "start":
      if (state === TimerState.IDLE || state === TimerState.PAUSED) return TimerState.RUNNING;
      return state; // RUNNING + start → no-op
    case "stop":
      if (state === TimerState.RUNNING) return TimerState.PAUSED;
      return state; // IDLE/PAUSED + stop → no-op
    case "reset":
      if (state === TimerState.RUNNING || state === TimerState.PAUSED) return TimerState.IDLE;
      return state; // IDLE + reset → no-op
    default:
      return state;
  }
}

/**
 * Format a millisecond duration as "MM:SS" (zero-padded).
 *
 * @param {number} ms - Duration in milliseconds (non-negative)
 * @returns {string}
 */
export function formatCountdown(ms) {
  const totalSeconds = Math.floor(Math.max(0, ms) / 1000);
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  return `${String(minutes).padStart(2, "0")}:${String(seconds).padStart(2, "0")}`;
}

/**
 * Validate a proposed Pomodoro duration.
 *
 * @param {number} n - Duration in whole minutes
 * @returns {boolean} true iff 1 ≤ n ≤ 120
 */
export function isValidDuration(n) {
  return Number.isInteger(n) && n >= 1 && n <= 120;
}

/**
 * Simulate a tick on a remainingMs value.
 * Decrements by 1000 and clamps to 0.
 *
 * @param {number} remainingMs
 * @returns {number}
 */
export function tickMs(remainingMs) {
  return Math.max(0, remainingMs - 1000);
}

// ============================================================
// --- Todo widget ---
// ============================================================

/**
 * Add a new todo task to the list.
 *
 * @param {string} label
 * @param {Array<{id: string, label: string, completed: boolean}>} tasks
 * @returns {Array<{id: string, label: string, completed: boolean}>}
 */
export function addTodo(label, tasks) {
  const id = `todo-${Date.now()}-${Math.random().toString(36).slice(2)}`;
  return [...tasks, { id, label, completed: false }];
}

/**
 * Toggle the completed state of a task by id.
 *
 * @param {string} id
 * @param {Array<{id: string, label: string, completed: boolean}>} tasks
 * @returns {Array<{id: string, label: string, completed: boolean}>}
 */
export function toggleTodo(id, tasks) {
  return tasks.map(task =>
    task.id === id ? { ...task, completed: !task.completed } : task
  );
}

/**
 * Delete a task by id.
 *
 * @param {string} id
 * @param {Array<{id: string, label: string, completed: boolean}>} tasks
 * @returns {Array<{id: string, label: string, completed: boolean}>}
 */
export function deleteTodo(id, tasks) {
  return tasks.filter(task => task.id !== id);
}

File: CodingCamp-20Apr26-RizkiPutraRamadhan-main/tests/pure_functions.py
"""
Pure functions mirroring the logic in js/app.js.
These must stay in sync with the JavaScript implementations.
"""

import math
import random
import time


# ============================================================
# Theme controller (pure logic)
# ============================================================

def toggle_theme_pure(theme: str) -> str:
    """
    Return the opposite theme.
    toggle_theme_pure(toggle_theme_pure(theme)) == theme  (involution)
    """
    return "dark" if theme == "light" else "light"


# ============================================================
# Greeting widget
# ============================================================

def get_greeting_phrase(hour: int) -> str:
    """
    Map an hour (0–23) to the appropriate time-of-day greeting.

    Mapping:
      5–11  -> "Good Morning"
      12–17 -> "Good Afternoon"
      18–21 -> "Good Evening"
      0–4, 22–23 -> "Good Night"
    """
    if 5 <= hour <= 11:
        return "Good Morning"
    elif 12 <= hour <= 17:
        return "Good Afternoon"
    elif 18 <= hour <= 21:
        return "Good Evening"
    else:
        return "Good Night"


# ============================================================
# Timer widget
# ============================================================

TIMER_STATES = {"IDLE", "RUNNING", "PAUSED"}
TIMER_ACTIONS = {"start", "stop", "reset"}


def apply_timer_action(state: str, action: str) -> str:
    """
    Pure state-machine transition function.

    Transition table:
      IDLE    + start -> RUNNING
      RUNNING + stop  -> PAUSED
      RUNNING + reset -> IDLE
      PAUSED  + start -> RUNNING
      PAUSED  + reset -> IDLE
      IDLE    + stop  -> IDLE   (no-op)
      IDLE    + reset -> IDLE   (no-op)
      PAUSED  + stop  -> PAUSED (no-op)
      RUNNING + start -> RUNNING (no-op)
    """
    if action == "start":
        if state in ("IDLE", "PAUSED"):
            return "RUNNING"
        return state
    elif action == "stop":
        if state == "RUNNING":
            return "PAUSED"
        return state
    elif action == "reset":
        if state in ("RUNNING", "PAUSED"):
            return "IDLE"
        return state
    return state


def format_countdown(ms: int) -> str:
    """
    Format a millisecond duration as "MM:SS" (zero-padded).
    e.g. 1_500_000 ms -> "25:00", 65_000 ms -> "01:05"
    """
    total_seconds = math.floor(max(0, ms) / 1000)
    minutes = total_seconds // 60
    seconds = total_seconds % 60
    return f"{minutes:02d}:{seconds:02d}"


def is_valid_duration(n) -> bool:
    """
    Validate a proposed Pomodoro duration.
    Returns True iff n is an integer and 1 <= n <= 120.
    """
    return isinstance(n, int) and not isinstance(n, bool) and 1 <= n <= 120


def tick_ms(remaining_ms: int) -> int:
    """
    Simulate one tick: decrement by 1000 and clamp to 0.
    """
    return max(0, remaining_ms - 1000)


# ============================================================
# Todo widget
# ============================================================

def add_todo(label: str, tasks: list) -> list:
    """
    Add a new todo task to the list.
    Returns a new list with the task appended.
    """
    new_id = f"todo-{int(time.time() * 1000)}-{random.randint(0, 999999):06d}"
    return tasks + [{"id": new_id, "label": label, "completed": False}]


def toggle_todo(task_id: str, tasks: list) -> list:
    """
    Toggle the completed state of a task by id.
    Returns a new list.
    """
    return [
        {**t, "completed": not t["completed"]} if t["id"] == task_id else t
        for t in tasks
    ]


def delete_todo(task_id: str, tasks: list) -> list:
    """
    Delete a task by id.
    Returns a new list without the matching task.
    """
    return [t for t in tasks if t["id"] != task_id]
